### PR TITLE
#10258 Minimum Ajax load animation duration

### LINKED ISSUE #10258: CommandButton/etc: ajaxLoadingIndicator should have a minimum display time
Currently if you click on it, and the request e.g. just takes 50ms, you just see a weird flickering

what about setting a minimum display like 500ms?

- [ ] Minimum display so it shows either for the minimum duration or the full duration if it is longer than the minimum
- [ ] Delay duration where say a 100ms delay is set it won't show the spinner until 100ms and checks if `AjaxQueue.isEmpty() to see if the AJAX already finished to prevent short displays or flickering.

### PR DESCRIPTION
**Pull request by jepsar**
Fixes #10258

I was not sure how and where to define the minimum number of milliseconds to show Ajax load animations.

### CODE REVIEW

File: primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js [hunk 1]
```diff
@@ -421,19 +421,29 @@
...
-                    button.removeClass('ui-state-loading');
-
-                    if (widget.hasOwnProperty('enable')
-                        && widget.cfg.disableOnAjax !== false
-                        && !widget.cfg.disabledAttr) {
-                        widget.enable();
-                    }
-
-                    button.find('.ui-icon-loading').remove();
+                    setTimeout(PrimeFaces.buttonEndAjaxDisabled,
```
Comment (melloware) [inline primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js]:
```suggestion
                    PrimeFaces.queueTask(PrimeFaces.buttonEndAjaxDisabled,
```

**Review by melloware — CHANGES_REQUESTED** [review]

File: primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js [hunk 2]
```diff
@@ -400,9 +400,9 @@
                     }
 
                     button.addClass('ui-state-loading');
+                    button.data('astart', Date.now());
```
Comment (melloware) [inline primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js]: Should `astart` be a more recognizable name like `ajaxstart` or something?

**Comment by jepsar** [comment]
This should also fix a regression introduced with #10999

**Review by melloware — CHANGES_REQUESTED** [review]

File: primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js (hunk 1 above)
Comment (jepsar) [reply primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js]: How can you pass parameters using this? I'm now getting `Uncaught TypeError: Cannot read properties of undefined (reading 'removeClass')`

Comment (jepsar) [reply primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js]: Alright, I think I've got it... @melloware please have a look

**Review by melloware — CHANGES_REQUESTED** [review]
This looks much better!  Looks like TypScript issue with args

```
TypedError: [tagOptionalForRest] Parameter args was marked as optional, but code specifies this as a rest parameter (...args) and such arguments cannot be optional.
```

**Comment by tandraschko** [comment]
Build failing

I will review next week

**Comment by jepsar** [comment]
I think the minimum delay (of 500ms) is causing some tests to fail. I think I'll need to update those tests. Executing this JS before clicking buttons should fix it.
```js
PrimeFaces.ajax.minLoadAnim = 0;
```

**Comment by melloware** [comment]
@jepsar in the Integration tests I have 1 spot where I disableAnimations you could just put the code there to lower the timeout.

**Comment by tandraschko** [comment]
Why does it cause failing tests?
Do we have tests which verifies the loading indicator

**Comment by jepsar** [comment]
> Why does it cause failing tests? Do we have tests which verifies the loading indicator

ConfirmDialog tests checked if buttons were enabled while they were not (yet) due to the minimum loading time.

I'll add a test for the loading indicator.

**Comment by jepsar** [comment]
> @jepsar in the Integration tests I have 1 spot where I disableAnimations you could just put the code there to lower the timeout.

This is not really an animation, so I don't think it should be controlled by that function.

**Comment by jepsar** [comment]
@melloware WDYT, show we mention `PrimeFaces.ajax.minLoadAnim` in the documentation somewhere?

File: primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.commandlink.js [hunk 3]
```diff
@@ -56,6 +57,7 @@ PrimeFaces.widget.CommandLink = PrimeFaces.widget.BaseWidget.extend({
                 }
                 if (PrimeFaces.ajax.Utils.isXhrSource($this, settings)) {
                     $this.jq.addClass('ui-state-loading');
+                    $this.ajaxStart = Date.now();
```
Comment (melloware) [inline primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.commandlink.js:60]: Should this value be cleared somewhere once its no longer used so the value in memory is not misleading?

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by melloware** [comment]
I feel like in the AJAX section you could mention it somewhere like here: https://primefaces.github.io/primefaces/13_0_0/#/core/ajaxJavascriptApi

**Review by melloware — APPROVED** [review]

**Comment by jepsar** [comment]
> I feel like in the AJAX section you could mention it somewhere like here: https://primefaces.github.io/primefaces/13_0_0/#/core/ajaxJavascriptApi

I also need to add this to what's new!

File: primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/confirmdialog/ConfirmDialog001Test.java [hunk 4]
```diff
@@ -47,6 +47,7 @@ class ConfirmDialog001Test extends AbstractPrimePageTest {
     void showDialog(Page page) {
         // Arrange
         ConfirmDialog dialog = page.dialog;
+        noMinLoadAnim();
```
Comment (tandraschko) [inline primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/confirmdialog/ConfirmDialog001Test.java]: Why is this required? Makes the tests a bit hacky

File: primefaces/src/main/resources/META-INF/resources/primefaces/core/core.ajax.js [hunk 5]
```diff
@@ -86,6 +86,12 @@ if (!PrimeFaces.ajax) {
             'onco': 'oncomplete'
         },
 
+        /**
+         * Minimum number of milliseconds to show Ajax load animations.
+         * @type {number}
+         */
+        minLoadAnim : 500,
```
Comment (tandraschko) [inline primefaces/src/main/resources/META-INF/resources/primefaces/core/core.ajax.js]: lets use "Animation" instead of "Anim", no need to shorten it actually
min/max is common

File: primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/confirmdialog/ConfirmDialog001Test.java (hunk 4 above)
Comment (jepsar) [reply primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/confirmdialog/ConfirmDialog001Test.java]: ConfirmDialog tests checked if buttons were enabled while they were not (yet) due to the minimum loading time.

Comment (tandraschko) [reply primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/confirmdialog/ConfirmDialog001Test.java]: you mean the assertClickable(page.confirm);?

maybe we should instead improve this and check the animination state and wait for it

Comment (jepsar) [reply primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/confirmdialog/ConfirmDialog001Test.java]: Alright. I'll also add some additional tests to the IT I've added.

File: primefaces/src/main/resources/META-INF/resources/primefaces/core/core.ajax.js (hunk 5 above)
Comment (jepsar) [reply primefaces/src/main/resources/META-INF/resources/primefaces/core/core.ajax.js]: Will do.

**Comment by melloware** [comment]
merged my Confirm Dialog changes.  @jepsar when you test locally can you uncomment out the two lines in the test 

```java
        // Assert
        //PrimeSelenium.waitGui().until(PrimeExpectedConditions.visibleInViewport(page.message));
        //assertEquals("Full page submitted", page.message.getMessage(0).getDetail());
```

This passes for me fine locally but fails on GitHub actions I just want to see if it passes for you also.

**Comment by jepsar** [comment]
> This passes for me fine locally but fails on GitHub actions I just want to see if it passes for you also.

@melloware Yeah, all tests work for me locally.

**Comment by melloware** [comment]
OK comment those lines back out please this may be a debug job for @christophs78 !

**Review by melloware — APPROVED** [review]

**Comment by jepsar** [comment]
I don't get why it's failing on one instance. Race condition?

**Comment by christophs78** [comment]
Sometimes it helps to add something like `@RepeatedTest(100)` to make a issue kind of reproduceable.
Additional we create since a few days screenshots on Github Actions whenever a IT fails.
image

I´ll try to look into into. Maybe a have to continue next week when it takes more time.

**Comment by christophs78** [comment]
> merged my Confirm Dialog changes. @jepsar when you test locally can you uncomment out the two lines in the test
> 
> ```java
>         // Assert
>         //PrimeSelenium.waitGui().until(PrimeExpectedConditions.visibleInViewport(page.message));
>         //assertEquals("Full page submitted", page.message.getMessage(0).getDetail());
> ```
> 
> This passes for me fine locally but fails on GitHub actions I just want to see if it passes for you also.

@melloware 
Your fix for https://github.com/primefaces/primefaces/issues/99, PR https://github.com/primefaces/primefaces/pull/11228 works only with CSP disabled. (CI enables CSP, Showcase has it disabled OOTB)
Spent a bit too much time on it because i looked into other things first. But thats the reason.

**Comment by christophs78** [comment]
> I don't get why it's failing on one instance. Race condition? Maybe invert the if conditions...

Looking into this the next few days.

**Comment by melloware** [comment]
Thanks @christophs78 i will debug with CSP on.

**Comment by melloware** [comment]
@jepsar i just updated your branch with my latest change that fix the tests for my issue #99
